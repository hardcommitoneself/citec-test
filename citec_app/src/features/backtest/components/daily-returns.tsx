// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { Ts } from 'features/utils/interfaces/backtest-api-resp'
import moment from 'moment'
import ReactApexChart from 'react-apexcharts'

interface DailyReturnsProps {
   ts_backtest: Ts
}

export const DailyReturns = ({ ts_backtest }: DailyReturnsProps) => {
   const { benchmark, portfolio, date } = ts_backtest

   const portfolioConverted = portfolio.map((num) => Number((num * 100).toFixed(3)))
   const benchmarkConverted = benchmark.map((num) => Number((num * 100).toFixed(3)))
   // Y AXIS

   const series = [
      {
         name: 'Portfolio',
         data: portfolioConverted,
      },
      {
         name: 'Benchmark',
         data: benchmarkConverted,
      },
   ]

   // const type = "line" as const;

   const options: ApexCharts.ApexOptions = {
      chart: {
         type: 'line' as const,
         height: 350,
         background: '#f6f7fb',
      },
      dataLabels: {
         enabled: false,
      },
      labels: date,
      stroke: {
         curve: 'straight' as const,
         width: 1.5,
      },
      grid: {
         row: {
            colors: ['#f3f3f3', 'transparent'], // takes an array which will be repeated on columns
            opacity: 0.5,
         },
      },
      xaxis: {
         tickAmount: 11,
         categories: [],
         labels: {
            rotate: 0,
            formatter(value) {
               const formatedDate = moment(value, 'DD-MM-YYYY').format('MM-YYYY')

               return formatedDate
            },
         },
      },
      yaxis: {
         tickAmount: undefined, // apexcharts.com/docs/options/yaxis/
         floating: false,
         decimalsInFloat: 0,
      },
   }

   return (
      <ReactApexChart options={options} series={series} type="line" height={350} />
   )
}
