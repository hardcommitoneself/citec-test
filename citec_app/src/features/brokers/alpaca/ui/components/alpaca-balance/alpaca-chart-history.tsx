// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import ReactApexChart from 'react-apexcharts'
import { formatToMoney } from 'features/utils/format-to-money'

interface HistoryProps {
   data: number[][]
}

export const ChartHistoryPortfolio = (data: HistoryProps) => {
   const series = [
      {
         name: '',
         data: data.data,
      },
   ]

   const options = {
      colors: ['#21BAB7'],
      stroke: {
         width: 2,
         curve: 'straight',
      },
      chart: {
         id: 'area-datetime',
         type: 'area',
         height: 350,
         toolbar: {
            show: false,
         },
         zoom: {
            autoScaleYaxis: true,
         },
         animations: {
            enabled: false,
         },
      },
      dataLabels: {
         enabled: false,
      },
      markers: {
         size: 0,
         style: 'hollow',
      },
      xaxis: {
         type: 'datetime',
         tooltip: {
            enabled: false,
         },
         labels: {
            datetimeFormatter: {
               year: 'MM/dd/yyyy',
               month: 'MM/dd/yyyy',
               day: 'MM/dd/yyyy',
               hour: 'HH:mm'
            }
         }
      },
      yaxis: {
         labels: {
            formatter: function (value: any) {
               return formatToMoney(value)
            },
         },
      },
      tooltip: {
         x: {
            format: 'dd MMM yyyy',
         },
      },
      fill: {
         type: 'gradient',
         gradient: {
            shadeIntensity: 1,
            opacityFrom: 0.7,
            opacityTo: 0.9,
            stops: [0, 100],
         },
      },
   }

   return (
      <div id="chart">
         <div>
            <ReactApexChart
               options={options as never}
               series={series as any}
               type="area"
               height={350}
            />
         </div>
      </div>
   )
}
