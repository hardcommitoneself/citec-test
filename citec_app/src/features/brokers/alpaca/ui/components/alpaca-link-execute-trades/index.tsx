// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import {
   getAllStatus,
   getTickers,
   processBuy,
   removeCash,
} from 'features/brokers/alpaca/ui/components/alpaca-link-execute-trades/execute-trades-utils'
import { TradeAlpaca } from 'features/brokers/alpaca/utils/trade-alpaca'
import { cancelPosition } from 'features/brokers/alpaca/utils/cancel-position-alpaca'
import { useAppSelector } from 'store/hooks'
import { Button } from '@mui/material'
import { executeOperations } from 'features/brokers/alpaca/ui/components/alpaca-link-execute-trades/execute-trades-utils'

import Swal from 'sweetalert2'
import { useAppNavigate } from 'features/utils/hooks'
import { validateAssetsExchange } from 'features/brokers/utils/validate-assets-exchange'
import { messageHighMape } from 'features/brokers/utils/message-high-mape'


export type linkExecuteTradesProps = {
   data: any
   mape: number
}

export const LinkExecuteTrades = ({ data, mape }: linkExecuteTradesProps) => {
   const visibility = process.env.REACT_APP_ALPACA_VISIBILITY

   const navigate = useAppNavigate()

   const brokers = useAppSelector((state) => state.brokers) 

   const executeTrades = (data: any) => {

      const tickers = getTickers(removeCash(data))
      const isValid = validateAssetsExchange(brokers.accountTypeSelect, tickers)

      isValid.then((valid) => {
         if (valid) {
            Swal.fire({
               title: 'Execute trades',
               text: 'Are you sure you want to execute these trades?',
               icon: 'warning',
               showCancelButton: false,
               confirmButtonText: 'Yes, Execute it!',
               cancelButtonText: '<span style="color:#79828D">No, cancel!</span>',
               showLoaderOnConfirm: true,
               confirmButtonColor: '#9C2B85',
               reverseButtons: true,
            }).then((result) => {
               if (result.isConfirmed) {
                  Swal.fire({
                     title: 'Execute trades',
                     html: `Processing the orders:
                     <ul>                  
                        <li><b>Sell</b>: - / - trades</li>
                        <li><b>Buy</b>: - / - trades</li>
                     </ul>`,
                     allowOutsideClick: false,
                     didOpen: async () => {
                        Swal.showLoading()
                        const { sell, buy } = await executeOperations(data)
                        const totalBuy = buy.length
                        const totalSell = sell.length
                        let valBuy = 0
                        let valSell = 0
                        const ordersIds: string[] = []
      
                        sell.map(async (s) => {
                           const resultSell = await cancelPosition(
                              brokers.accountTypeSelect,
                              s.symbol
                           )
      
                           ordersIds.push(
                              resultSell !== undefined ? resultSell.data.id : 'error'
                           )
      
                           valSell++
                           Swal.update({
                              html: `Processing the orders:
                                    <ul>                                 
                                       <li><b>Sell</b>: ${valSell} / ${totalSell} trades</li>
                                       <li><b>Buy</b>: ${valBuy} / ${totalBuy} trades</li>
                                    </ul>`,
                           })
                        })
      
                        setTimeout(() => {
                           const statusOrders = getAllStatus(
                              brokers.accountTypeSelect,
                              ordersIds
                           )
      
                           const proceedToBuy = processBuy(statusOrders)
                           if (proceedToBuy) {
                              buy.map(async (b) => {
                                 await TradeAlpaca(brokers.accountTypeSelect, b)
                                 valBuy++
                                 Swal.update({
                                    html: `Processing the orders:
                                          <ul>                                    
                                             <li><b>Sell</b>: ${valSell} / ${totalSell} trades</li>
                                             <li><b>Buy</b>: ${valBuy} / ${totalBuy} trades</li>
                                          </ul>`,
                                 })
                              })
                           }
                        }, 1000)
      
                        Swal.hideLoading()
                        Swal.update({
                           showCancelButton: true,
                           showConfirmButton: false,
                           cancelButtonText: '<span style="color:#79828D">Close</span>',
                           cancelButtonColor: '#E8EFF6',
                        })
                        navigate('/')
                     },
                  })
               }
            })
         } else {
            Swal.fire({
               title: "Trades can't be executed",
               text: "Trades can't be executed because at least one of the Alpaca market assets does not match",
               icon: 'warning',
               showCancelButton: false,
               confirmButtonText: 'Ok!',
               confirmButtonColor: '#9C2B85',
               allowOutsideClick: true,
               backdrop: true
            })
         }
      })
      
   }

   if (mape > 5) {
      messageHighMape(mape)
   }

   return (
      <>
         {visibility !== 'undefined' && visibility == 'true' && (
            <div className='inline mr-3.5'>
               <Button
                  variant="actionButtonCitec"
                  className= "h-[35px] min-w-0"
                  onClick={() => executeTrades(data)}
                  data-testid="test-execute-trades"
               >
                  Execute trades
               </Button>
            </div>
            
         )}
      </>
   )
}
