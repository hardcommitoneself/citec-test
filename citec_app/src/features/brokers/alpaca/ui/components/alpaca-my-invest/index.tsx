// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { useState, useEffect } from 'react'
import { Button } from '@mui/material'
import { useAppSelector } from 'store/hooks'
import { InvestTable } from 'features/brokers/utils/table/invest-table'
import { columns } from 'features/brokers/utils/table/my-invest'
import {
   getStockForValidation,
   buildDataInvest,
} from 'features/brokers/alpaca/utils/positions-alpaca-utils'
import { useDispatch } from 'react-redux'
import { setPortfolio } from 'store/slices/portfolio'
import { setWorkflow } from 'store/slices/workflow'
import { PositionAlpaca } from 'features/brokers/alpaca/utils/position-alpaca'
import { getAssetType } from 'features/brokers/utils/get-asset-type'
import { formatInvestmentData } from 'features/brokers/alpaca/utils/format-investment-data'
import XLSX from 'xlsx'
import { useAppNavigate } from 'features/utils/hooks'
import { useSelectUniverse } from 'features/utils/hooks'


export interface MyInvestments {
   asset_class: string
   quantity: number
   price: number
   investment: number
   day_gain_loss: number[]
   total_gain_loss: number[]
   url_logo?: string
   subRows: any
}

export let loadPortfolioOnRedux: any
export let handleOnExportInvestments: any

export const AlpacaMyInvest = () => {
   const visibility = process.env.REACT_APP_ALPACA_VISIBILITY

   const brokers = useAppSelector((state) => state.brokers)

   const [dataTable, setDataTable] = useState<MyInvestments[]>([])

   const navigate = useAppNavigate()
   const dispatch = useDispatch()

   const { handleSelectBenckmark } = useSelectUniverse()

   const dataInvest = async () => {
      const positions = await PositionAlpaca(brokers.accountTypeSelect)

      const stocks = getStockForValidation(positions)
      const assetType = await getAssetType(stocks)
      const response = buildDataInvest(positions, assetType)
      setDataTable(response)
   }

   useEffect(() => {
      dataInvest()
   }, [brokers.accountTypeSelect])

   loadPortfolioOnRedux = (row: any, flow: string) => {

      if (flow === 'rebalance') {
         dispatch(setWorkflow('rebalance'))
      } else {
         dispatch(setWorkflow('load'))
      }

      const ticker = []
      const outstanding_balance = []
      const unrealized_capital_gains = []
      for (let i = 0; i < row.length; i++) {
         ticker.push(row[i].values.asset_class)
         outstanding_balance.push(Number(row[i].values.investment))
         unrealized_capital_gains.push(Number(row[i].values.total_gain_loss[0]))
      }
      dispatch(
         setPortfolio({
            ticker,
            outstanding_balance,
            unrealized_capital_gains,
         })
      )

      handleSelectBenckmark('SPY')

      if (flow === 'load') {
         navigate('/portfolio-loader')
      } else if (flow === 'analysis') {
         navigate('/portfolio-analysis')
      } else if (flow === 'rebalance') {
         navigate('/investment')
      }
   }

   handleOnExportInvestments = (row: any) => {
      try {
         const data_investment = []
         data_investment.push(formatInvestmentData(row.values))

         for (const i of row.subRows) {
            data_investment.push(formatInvestmentData(i.values))
         }

         const wordBook = XLSX.utils.book_new()
         const wordSheet = XLSX.utils.json_to_sheet(data_investment)

         XLSX.utils.book_append_sheet(wordBook, wordSheet)
         XLSX.writeFile(wordBook, 'my_investments.xlsx')

         return 1
      } catch (error) {
         return error
      }
   }

   return (
      <>
         {visibility == 'true' && (
            <div className="mx-10 mt-8 bg-white p-4 rounded-2xl relative balance">
               <div className="flex buying-power">
                  <h3 className="font-semibold text-[26px] text-[#2d405a]">
                     My Investments
                  </h3>
               </div>
               <InvestTable columns={columns} data={dataTable} messageNoData="" />
            </div>
         )}
      </>
   )
}
