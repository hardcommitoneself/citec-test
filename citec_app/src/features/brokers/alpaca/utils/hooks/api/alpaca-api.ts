// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import axios from 'axios'
import { Security } from 'features/utils/security'

const baseURLPaper = process.env.REACT_APP_ALPACA_URL_PAPER
const baseURLLive = process.env.REACT_APP_ALPACA_URL_LIVE

type FetchDataFuncProps = {
   type: string
   url: string
   method: 'GET' | 'POST' | 'DELETE' | 'PUT'
   payload?: any
   params?: any
}

export const alpacaApiConnect = axios.create({
   headers: {
      accept: 'application/json',
   },
   timeout: 1000 * 40,
})

export const AlpacaApi = (DataFuncProps: FetchDataFuncProps) => {
   const tokenAlpaca = Security.decrypt(String(localStorage.getItem('a1c_t')))

   let baseURL
   if (DataFuncProps.type == 'PAPER') baseURL = baseURLPaper
   else if (DataFuncProps.type == 'LIVE') baseURL = baseURLLive

   alpacaApiConnect.defaults.headers.common[
      'Authorization'
   ] = `Bearer ${tokenAlpaca}`

   return alpacaApiConnect.request({
      method: DataFuncProps.method,
      baseURL,
      url: DataFuncProps.url,
      data: DataFuncProps.payload,
      params: DataFuncProps.params,
   })
}
