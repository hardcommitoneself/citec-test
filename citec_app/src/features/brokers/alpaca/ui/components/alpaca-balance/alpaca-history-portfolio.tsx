// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { useState, useEffect } from 'react'
import { ChartHistoryPortfolio } from './alpaca-chart-history'
import { AlpacaHistory } from 'features/brokers/alpaca/utils/alpaca-history'
import { useAppSelector } from 'store/hooks'
import { formatToMoney } from 'features/utils/format-to-money'
import { numberToFixed } from 'features/utils/to-fixed-number'

export const HistoryPortfolio = (/*Propiedades*/) => {
   const [selection, setSelection] = useState('1D')
   const [dataHistory, setDataHistory] = useState([[0, 0]])
   const [equity, setEquity] = useState(0)
   const [profitLoss, setProfitLoss] = useState(0)
   const [profitLossPer, setProfitLossPer] = useState(0)

   const brokers = useAppSelector((state) => state.brokers)

   const buttons = ['1D', '1W', '1M', '3M', '1Y', 'YTD']

   useEffect(() => {
      setDataHistoryBalance(selection)

      const interval = setInterval(() => {
         setDataHistoryBalance(selection)
      }, 5000)

      return function cleanup() {
         clearInterval(interval)
      }
   }, [selection, brokers.accountTypeSelect])

   const setDataHistoryBalance = (timeline: string) => {
      AlpacaHistory(
         brokers.accountTypeSelect,
         timeline,
         (
            dataHistory: number[][],
            actualEquity: number,
            profitLoss: number,
            profitLossPer: number
         ) => {
            setDataHistory(dataHistory)
            setEquity(actualEquity)
            setProfitLoss(profitLoss)
            setProfitLossPer(profitLossPer)
         },
         (error: any) => {
            if (error.response) {
               console.log(error.response.data)
               console.log(error.response.status)
               console.log(error.response.headers)
            }
         }
      )
   }

   const updateData = (timeline: string) => {
      setSelection(timeline)
      setDataHistoryBalance(timeline)
   }

   return (
      <div>
         <div className="head-broker">
            <div className="bodyBroker">
               <img src={require('assets/img/brokers/broker-balance.png')} />
               <div className="data-head-invest">
                  <span className="label-head-invest">My Investments</span>
                  <span className="amount-balance">
                     {formatToMoney(Math.trunc(equity))}
                  </span>
               </div>
               <div
                  className={
                     profitLoss >= 0 ? 'profitloss profit' : 'profitloss loss'
                  }
               >
                  {profitLoss >= 0 ? '+' : ''}
                  {formatToMoney(profitLoss)} ({numberToFixed(profitLossPer)}%)
               </div>
            </div>
            <div className="toolbar-hitory">
               {buttons.map((btnHistory) => (
                  <button
                     key={btnHistory}
                     id={btnHistory}
                     onClick={() => updateData(btnHistory)}
                     className={
                        selection === btnHistory
                           ? 'btn-filter-history active'
                           : 'btn-filter-history'
                     }
                  >
                     {btnHistory}
                  </button>
               ))}
            </div>
         </div>
         <div id="chart">
            <ChartHistoryPortfolio data={dataHistory} />
         </div>
      </div>
   )
}
