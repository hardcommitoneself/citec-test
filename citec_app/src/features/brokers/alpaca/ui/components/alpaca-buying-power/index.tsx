// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { useState, useEffect } from 'react'
import { Button } from '@mui/material'
import { formatToMoney } from 'features/utils/format-to-money'
import { AlpacaApi } from 'features/brokers/alpaca/utils/hooks/api/alpaca-api'
import { useDispatch } from 'react-redux'
import { resetBrokers } from 'store/slices/brokers'
import { useAppSelector } from 'store/hooks'

export const AlpacaBuyingPower = () => {
   const dispatch = useDispatch()

   const visibility = process.env.REACT_APP_ALPACA_VISIBILITY
   const [buyingPower, setBuyingPower] = useState(0)

   const brokers = useAppSelector((state) => state.brokers)

   useEffect(() => {
      getBuyinPower(brokers.accountTypeSelect)
      const interval = setInterval(() => {
         getBuyinPower(brokers.accountTypeSelect)
      }, 5000)

      return function cleanup() {
         clearInterval(interval)
      }
   }, [brokers.accountTypeSelect])

   const getBuyinPower = (type: string) => {
      AlpacaApi({
         type: type,
         url: 'account',
         method: 'GET',
      })
         .then(function (response) {
            setBuyingPower(response.data.buying_power)
         })
         .catch(function (error) {
            if (error.response) {
               if (error.response.status == '403') {
                  console.log(error.response.status)
                  dispatch(resetBrokers())
                  localStorage.removeItem('a1c_t')
               } else {
                  console.log(error.response.data)
                  console.log(error.response.status)
                  console.log(error.response.headers)
               }
            }
         })
   }

   return (
      <>
         {visibility == 'true' && (
            <div className="mx-10 mt-8 bg-white p-4 rounded-2xl relative balance">
               <div className="flex buying-power">
                  <h3 className="font-semibold text-[26px] text-[#2d405a]">
                     Buying Power
                  </h3>
                  <h3 className="font-semibold text-[26px] text-[#2d405a] buying-power-value">
                     {formatToMoney(Math.trunc(buyingPower))}
                  </h3>
               </div>
            </div>
         )}
      </>
   )
}
