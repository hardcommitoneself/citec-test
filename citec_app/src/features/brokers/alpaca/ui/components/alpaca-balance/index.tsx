// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { useDispatch } from 'react-redux'
import { LinkAccount } from 'features/brokers/link-account'
import MenuItem from '@mui/material/MenuItem'
import FormControl from '@mui/material/FormControl'
import Select, { SelectChangeEvent } from '@mui/material/Select'
import { setAccountTypeSelect } from 'store/slices/brokers'
import { useAppSelector } from 'store/hooks'
import { HistoryPortfolio } from './alpaca-history-portfolio'

export const AlpacaBalance = () => {
   const visibility = process.env.REACT_APP_ALPACA_VISIBILITY

   const dispatch = useDispatch()

   const brokers = useAppSelector((state) => state.brokers)
   const typesAccount = brokers.accountType

   const handleChange = (event: SelectChangeEvent) => {
      const typeSelected = event.target.value as string
      dispatch(setAccountTypeSelect({ accountTypeSelect: typeSelected }))
   }

   return (
      <>
         {visibility == 'true' && (
            <div className="mx-10 mt-2 bg-white p-4 rounded-2xl relative balance">
               <div className="brokerTitle">
                  <div className="actionBroker">
                     <div className="selectAccount">
                        <span className="text-sm">Account</span>
                        <FormControl
                           variant="standard"
                           sx={{ m: 1, width: 200, pt: 0 }}
                           size="small"
                        >
                           <Select
                              value={brokers.accountTypeSelect}
                              label="type-acount"
                              onChange={handleChange}
                              inputProps={{ 'aria-label': 'Without label' }}
                              className="selectAccountType"
                           >
                              {typesAccount.map((typeAc) => (
                                 <MenuItem value={typeAc} key={typeAc}>
                                    {typeAc == 'PAPER'
                                       ? 'Paper Trading'
                                       : 'Live Trading'}
                                 </MenuItem>
                              ))}
                           </Select>
                        </FormControl>
                     </div>

                     <LinkAccount />
                  </div>
               </div>
               <HistoryPortfolio />
            </div>
         )}
      </>
   )
}
