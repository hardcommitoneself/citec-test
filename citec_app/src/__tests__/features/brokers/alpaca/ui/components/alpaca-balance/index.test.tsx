// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import '@testing-library/jest-dom'
import { render, screen, prettyDOM } from '@testing-library/react'
import { AlpacaBalance } from 'features/brokers/alpaca/ui/components/alpaca-balance'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { alpacaApiConnect } from 'features/brokers/alpaca/utils/hooks/api/alpaca-api'
import MockAdapter from 'axios-mock-adapter'

process.env.REACT_APP_ALPACA_VISIBILITY = 'true'

jest.mock('react-apexcharts', () => {
   return {
      __esModule: true,
      default: () => {
         return <div />
      },
   }
})

const mockAlpaca = new MockAdapter(alpacaApiConnect)

const initialState = {
   brokers: {
      brokerConnected: 'Alpaca',
      balance: 0,
      accountType: ['PAPER', 'LIVE'],
      a1c_t: '',
      accountTypeSelect: 'PAPER',
      buyingPower: 10000,
   },
}
const mockStore = configureStore()
const store = mockStore(initialState)
describe('Brokers redux state tests', () => {
   test('renders Balance Alpaca Component', () => {
      jest.useFakeTimers()
      jest.spyOn(global, 'setInterval')

      const resp = {
         data: {
            timestamp: [1654300800, 1654560000, 1654646400, 1654732800, 1654808365],
            equity: [100419.38, 100673.44, 102367.46, 101341.29, 101075.86],
            profit_loss: [-1846.45, 266.06, 1694.02, -1025.17, -265.43],
            profit_loss_pct: [
               0, 0.0026494885748149, 0.019520952256154384, 0.009310849410734142,
               -0.0026191693435124027,
            ],
            base_value: 0,
            timeframe: '1D',
         },
         status: 200,
         statusText: 'OK',
         headers: {
            'content-length': '326',
            'content-type': 'application/json; charset=UTF-8',
         },
         config: {
            transitional: {
               silentJSONParsing: true,
               forcedJSONParsing: true,
               clarifyTimeoutError: false,
            },
            transformRequest: [null],
            transformResponse: [null],
            timeout: 0,
            xsrfCookieName: 'XSRF-TOKEN',
            xsrfHeaderName: 'X-XSRF-TOKEN',
            maxContentLength: -1,
            maxBodyLength: -1,
            headers: {
               Accept: 'application/json',
               Authorization: 'Bearer 640a9d05-7f79-4aea-87d2-6beaec40f9a8',
            },
            method: 'get',
            url: 'https://paper-api.alpaca.markets/v2/account/portfolio/history',
            params: {
               timeframe: '1D',
               date_end: '2022-06-09',
               extended_hours: true,
               period: '1W',
               date_start: '',
            },
         },
         request: {},
      }
      mockAlpaca.onGet('account/portfolio/history').reply(200, resp)

      const { container } = render(
         <Provider store={store}>
            <AlpacaBalance />
         </Provider>
      )
      const linkElement = screen.getAllByText(/Account/i)
      expect(linkElement[0]).toBeInTheDocument()
      expect(container).toMatchSnapshot()
      jest.useRealTimers()
   })
})
