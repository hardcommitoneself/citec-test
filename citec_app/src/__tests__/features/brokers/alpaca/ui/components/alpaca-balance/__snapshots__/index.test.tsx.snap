// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Brokers redux state tests renders Balance Alpaca Component 1`] = `
<div>
  <div
    class="mx-10 mt-2 bg-white p-4 rounded-2xl relative balance"
  >
    <div
      class="brokerTitle"
    >
      <div
        class="actionBroker"
      >
        <div
          class="selectAccount"
        >
          <span
            class="text-sm"
          >
            Account
          </span>
          <div
            class="MuiFormControl-root css-1vcie31-MuiFormControl-root"
          >
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall selectAccountType css-a3l6o-MuiInputBase-root-MuiInput-root-MuiSelect-root"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-label="Without label"
                class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jd1zyo-MuiSelect-select-MuiInputBase-input-MuiInput-input"
                role="button"
                tabindex="0"
              >
                Paper Trading
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="PAPER"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="linkAccount"
        >
          <button
            class="MuiButton-root MuiButton-actionButtonCitec MuiButton-actionButtonCitecPrimary MuiButton-sizeMedium MuiButton-actionButtonCitecSizeMedium MuiButtonBase-root  css-1w1rijm-MuiButtonBase-root-MuiButton-root"
            data-testid="btn-link-account"
            tabindex="0"
            type="button"
          >
            <span
              class="font-semibold"
            >
              + Link account
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="head-broker"
      >
        <div
          class="bodyBroker"
        >
          <img
            src="broker-balance.png"
          />
          <div
            class="data-head-invest"
          >
            <span
              class="label-head-invest"
            >
              My Investments
            </span>
            <span
              class="amount-balance"
            >
              $ 0
            </span>
          </div>
          <div
            class="profitloss profit"
          >
            +
            $ 0
             (
            0
            %)
          </div>
        </div>
        <div
          class="toolbar-hitory"
        >
          <button
            class="btn-filter-history active"
            id="1D"
          >
            1D
          </button>
          <button
            class="btn-filter-history"
            id="1W"
          >
            1W
          </button>
          <button
            class="btn-filter-history"
            id="1M"
          >
            1M
          </button>
          <button
            class="btn-filter-history"
            id="3M"
          >
            3M
          </button>
          <button
            class="btn-filter-history"
            id="1Y"
          >
            1Y
          </button>
          <button
            class="btn-filter-history"
            id="YTD"
          >
            YTD
          </button>
        </div>
      </div>
      <div
        id="chart"
      >
        <div
          id="chart"
        >
          <div>
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
